From 2e5398359097c9969eec0737a85a0fce09b27fdb Mon Sep 17 00:00:00 2001
From: Xihua Chen <xihua.chen@intel.com>
Date: Tue, 30 Apr 2019 16:06:26 +0800
Subject: [PATCH 1/2] Support ota update from udisk.

Add udisk entry in recovery menu and support ota from udisk.

Port from P of https://android.intel.com/#/c/631103/4

Change-Id: Id1f08bed9889f344a26aed8f9fdcbdde465d7071
Signed-off-by: Xihua Chen <xihua.chen@intel.com>
---
 install/Android.bp                           |   1 +
 install/fuse_udisk_install.cpp               | 215 +++++++++++++++++++
 install/include/install/fuse_udisk_install.h |  22 ++
 recovery.cpp                                 |  15 +-
 recovery_ui/device.cpp                       |   1 +
 recovery_ui/include/recovery_ui/device.h     |   1 +
 6 files changed, 253 insertions(+), 2 deletions(-)
 create mode 100644 install/fuse_udisk_install.cpp
 create mode 100644 install/include/install/fuse_udisk_install.h

diff --git a/install/Android.bp b/install/Android.bp
index ea893a07..015ba108 100644
--- a/install/Android.bp
+++ b/install/Android.bp
@@ -63,6 +63,7 @@ cc_library_static {
         "adb_install.cpp",
         "asn1_decoder.cpp",
         "fuse_sdcard_install.cpp",
+        "fuse_udisk_install.cpp",
         "install.cpp",
         "package.cpp",
         "verifier.cpp",
diff --git a/install/fuse_udisk_install.cpp b/install/fuse_udisk_install.cpp
new file mode 100644
index 00000000..ee1c46b7
--- /dev/null
+++ b/install/fuse_udisk_install.cpp
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "install/fuse_udisk_install.h"
+
+#include <dirent.h>
+#include <signal.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include <android-base/logging.h>
+#include <android-base/strings.h>
+
+#include "bootloader_message/bootloader_message.h"
+#include "fuse_provider.h"
+#include "fuse_sideload.h"
+#include "install/install.h"
+#include "otautil/roots.h"
+
+static constexpr const char* UDISKA_ROOT = "/udiska";
+static constexpr const char* UDISKB_ROOT = "/udiskb";
+// How long (in seconds) we wait for the fuse-provided package file to
+// appear, before timing out.
+static constexpr int UDISK_INSTALL_TIMEOUT = 10;
+
+// Set the BCB to reboot back into recovery (it won't resume the install from
+// udisk though).
+static void SetUdiskUpdateBootloaderMessage() {
+  std::vector<std::string> options;
+  std::string err;
+  if (!update_bootloader_message(options, &err)) {
+    LOG(ERROR) << "Failed to set BCB message: " << err;
+  }
+}
+
+// Returns the selected filename, or an empty string.
+static std::string BrowseDirectory(const std::string& path, Device* device, RecoveryUI* ui) {
+  ensure_path_mounted(path);
+
+  std::unique_ptr<DIR, decltype(&closedir)> d(opendir(path.c_str()), closedir);
+  if (!d) {
+    PLOG(ERROR) << "error opening " << path;
+    return "";
+  }
+
+  std::vector<std::string> dirs;
+  std::vector<std::string> entries{ "../" };  // "../" is always the first entry.
+
+  dirent* de;
+  while ((de = readdir(d.get())) != nullptr) {
+    std::string name(de->d_name);
+
+    if (de->d_type == DT_DIR) {
+      // Skip "." and ".." entries.
+      if (name == "." || name == "..") continue;
+      dirs.push_back(name + "/");
+    } else if (de->d_type == DT_REG && android::base::EndsWithIgnoreCase(name, ".zip")) {
+      entries.push_back(name);
+    }
+  }
+
+  std::sort(dirs.begin(), dirs.end());
+  std::sort(entries.begin(), entries.end());
+
+  // Append dirs to the entries list.
+  entries.insert(entries.end(), dirs.begin(), dirs.end());
+
+  std::vector<std::string> headers{ "Choose a package to install:", path };
+
+  size_t chosen_item = 0;
+  while (true) {
+    chosen_item = ui->ShowMenu(
+        headers, entries, chosen_item, true,
+        std::bind(&Device::HandleMenuKey, device, std::placeholders::_1, std::placeholders::_2));
+
+    // Return if WaitKey() was interrupted.
+    if (chosen_item == static_cast<size_t>(RecoveryUI::KeyError::INTERRUPTED)) {
+      return "";
+    }
+
+    const std::string& item = entries[chosen_item];
+    if (chosen_item == 0) {
+      // Go up but continue browsing (if the caller is BrowseDirectory).
+      return "";
+    }
+
+    std::string new_path = path + "/" + item;
+    if (new_path.back() == '/') {
+      // Recurse down into a subdirectory.
+      new_path.pop_back();
+      std::string result = BrowseDirectory(new_path, device, ui);
+      if (!result.empty()) return result;
+    } else {
+      // Selected a zip file: return the path to the caller.
+      return new_path;
+    }
+  }
+
+  // Unreachable.
+}
+
+static bool StartUdiskFuse(const std::string& path) {
+  auto file_data_reader = std::make_unique<FuseFileDataProvider>(path, 65536);
+
+  if (!file_data_reader->Valid()) {
+    return false;
+  }
+
+  // The installation process expects to find the udisk unmounted. Unmount it with MNT_DETACH so
+  // that our open file continues to work but new references see it as unmounted.
+  umount2("/udiska", MNT_DETACH);
+  umount2("/udiskb", MNT_DETACH);
+
+  return run_fuse_sideload(std::move(file_data_reader)) == 0;
+}
+
+int ApplyFromUdisk(Device* device, RecoveryUI* ui) {
+  const char* udisk_root;
+
+  udisk_root = UDISKA_ROOT;
+  if (ensure_path_mounted(udisk_root) != 0) {
+    LOG(ERROR) << "\n-- Couldn't mount " << udisk_root << ".\n";
+    udisk_root = UDISKB_ROOT;
+    if (ensure_path_mounted(udisk_root) != 0) {
+      LOG(ERROR) << "\n-- Couldn't mount " << udisk_root << ".\n";
+      return INSTALL_ERROR;
+    }
+  }
+
+  std::string path = BrowseDirectory(udisk_root, device, ui);
+  if (path.empty()) {
+    LOG(ERROR) << "\n-- No package file selected.\n";
+    ensure_path_unmounted(udisk_root);
+    return INSTALL_ERROR;
+  }
+
+  ui->Print("\n-- Install %s ...\n", path.c_str());
+  SetUdiskUpdateBootloaderMessage();
+
+  // We used to use fuse in a thread as opposed to a process. Since accessing
+  // through fuse involves going from kernel to userspace to kernel, it leads
+  // to deadlock when a page fault occurs. (Bug: 26313124)
+  pid_t child;
+  if ((child = fork()) == 0) {
+    bool status = StartUdiskFuse(path);
+
+    _exit(status ? EXIT_SUCCESS : EXIT_FAILURE);
+  }
+
+  // FUSE_SIDELOAD_HOST_PATHNAME will start to exist once the fuse in child
+  // process is ready.
+  int result = INSTALL_ERROR;
+  int status;
+  bool waited = false;
+  for (int i = 0; i < UDISK_INSTALL_TIMEOUT; ++i) {
+    if (waitpid(child, &status, WNOHANG) == -1) {
+      result = INSTALL_ERROR;
+      waited = true;
+      break;
+    }
+
+    struct stat sb;
+    if (stat(FUSE_SIDELOAD_HOST_PATHNAME, &sb) == -1) {
+      if (errno == ENOENT && i < UDISK_INSTALL_TIMEOUT - 1) {
+        sleep(1);
+        continue;
+      } else {
+        LOG(ERROR) << "Timed out waiting for the fuse-provided package.";
+        result = INSTALL_ERROR;
+        kill(child, SIGKILL);
+        break;
+      }
+    }
+
+    result = install_package(FUSE_SIDELOAD_HOST_PATHNAME, false, false, 0 /*retry_count*/, ui);
+    break;
+  }
+
+  if (!waited) {
+    // Calling stat() on this magic filename signals the fuse
+    // filesystem to shut down.
+    struct stat sb;
+    stat(FUSE_SIDELOAD_HOST_EXIT_PATHNAME, &sb);
+
+    waitpid(child, &status, 0);
+  }
+
+  if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+    LOG(ERROR) << "Error exit from the fuse process: " << WEXITSTATUS(status);
+  }
+
+  ensure_path_unmounted(udisk_root);
+  return result;
+}
diff --git a/install/include/install/fuse_udisk_install.h b/install/include/install/fuse_udisk_install.h
new file mode 100644
index 00000000..8a31f1c3
--- /dev/null
+++ b/install/include/install/fuse_udisk_install.h
@@ -0,0 +1,22 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "recovery_ui/device.h"
+#include "recovery_ui/ui.h"
+
+int ApplyFromUdisk(Device* device, RecoveryUI* ui);
diff --git a/recovery.cpp b/recovery.cpp
index 5fc673ec..88c2bee6 100644
--- a/recovery.cpp
+++ b/recovery.cpp
@@ -52,6 +52,7 @@
 #include "fsck_unshare_blocks.h"
 #include "install/adb_install.h"
 #include "install/fuse_sdcard_install.h"
+#include "install/fuse_udisk_install.h"
 #include "install/install.h"
 #include "install/package.h"
 #include "install/wipe_data.h"
@@ -543,6 +544,7 @@ static Device::BuiltinAction prompt_and_wait(Device* device, int status) {

       case Device::APPLY_ADB_SIDELOAD:
       case Device::APPLY_SDCARD:
+      case Device::APPLY_UDISK:
       case Device::ENTER_RESCUE: {
         save_current_log = true;

@@ -556,10 +558,19 @@ static Device::BuiltinAction prompt_and_wait(Device* device, int status) {
           status = ApplyFromAdb(device, false /* rescue_mode */, &reboot_action);
         } else {
           adb = false;
-          status = ApplyFromSdcard(device, ui);
+          if (chosen_action == Device::APPLY_SDCARD)
+            status = ApplyFromSdcard(device, ui);
+          else if (chosen_action == Device::APPLY_UDISK)
+            status = ApplyFromUdisk(device, ui);
         }

-        ui->Print("\nInstall from %s completed with status %d.\n", adb ? "ADB" : "SD card", status);
+        if (chosen_action == Device::APPLY_ADB_SIDELOAD)
+          ui->Print("\nInstall from ADB completed with status %d.\n", status);
+        else if (chosen_action == Device::APPLY_SDCARD)
+          ui->Print("\nInstall from SD card completed with status %d.\n", status);
+        else if (chosen_action == Device::APPLY_UDISK)
+          ui->Print("\nInstall from Udisk completed with status %d.\n", status);
+
         if (status == INSTALL_REBOOT) {
           return reboot_action;
         }
diff --git a/recovery_ui/device.cpp b/recovery_ui/device.cpp
index e7ae1a3e..6883bcb1 100644
--- a/recovery_ui/device.cpp
+++ b/recovery_ui/device.cpp
@@ -31,6 +31,7 @@ static std::vector<std::pair<std::string, Device::BuiltinAction>> g_menu_actions
   { "Enter fastboot", Device::ENTER_FASTBOOT },
   { "Apply update from ADB", Device::APPLY_ADB_SIDELOAD },
   { "Apply update from SD card", Device::APPLY_SDCARD },
+  { "Apply update from Udisk", Device::APPLY_UDISK },
   { "Wipe data/factory reset", Device::WIPE_DATA },
   { "Wipe cache partition", Device::WIPE_CACHE },
   { "Mount /system", Device::MOUNT_SYSTEM },
diff --git a/recovery_ui/include/recovery_ui/device.h b/recovery_ui/include/recovery_ui/device.h
index 7c76cdb0..980c3d99 100644
--- a/recovery_ui/include/recovery_ui/device.h
+++ b/recovery_ui/include/recovery_ui/device.h
@@ -58,6 +58,7 @@ class Device {
     REBOOT_FASTBOOT = 17,
     REBOOT_RECOVERY = 18,
     REBOOT_RESCUE = 19,
+    APPLY_UDISK = 20,
   };

   explicit Device(RecoveryUI* ui);
--
2.21.0
